File: player/atarist/src/state.c
/* state.c -- état runtime, évaluation des conditions, application des effets. */

#include "state.h"
#include "story.h"
#include "snd.h"

u8 stat_val[MAX_STATS];
u8 item_bits[(MAX_ITEMS + 7) / 8];
u8 flag_bits[(MAX_FLAGS + 7) / 8];
u16 g_score;
u16 g_moves;

void state_init(void)
{
    u8 i;
    for (i = 0; i < g_nstats; ++i) {
        stat_val[i] = stat_init[i];
        stat_max[i] = stat_maxdef[i];   /* max courant = defaut (le perso peut le changer) */
    }
    for (i = 0; i < (g_nitems + 7) / 8; ++i)
        item_bits[i] = item_default[i];
    for (i = 0; i < (g_nflags + 7) / 8; ++i)
        flag_bits[i] = flag_default[i];
    g_score = 0;
    g_moves = 0;
}

u8 item_get(u8 i) { return (item_bits[i >> 3] >> (i & 7)) & 1; }
u8 flag_get(u8 i) { return (flag_bits[i >> 3] >> (i & 7)) & 1; }

void state_clear_locals(void)
{
    u8 i;
    for (i = g_local_base; i < g_nflags; ++i)
        flag_bits[i >> 3] &= (u8)~(1u << (i & 7));
}

/* Somme d'un modificateur de combat sur tous les objets portes. */
int gear_atk(void)
{
    u8 i; int s = 0;
    for (i = 0; i < g_nitems; ++i) if (item_get(i)) s += item_atk[i];
    return s;
}
int gear_dmg(void)
{
    u8 i; int s = 0;
    for (i = 0; i < g_nitems; ++i) if (item_get(i)) s += item_dmg[i];
    return s;
}
int gear_armor(void)
{
    u8 i; int s = 0;
    for (i = 0; i < g_nitems; ++i) if (item_get(i)) s += item_armor[i];
    return s;
}

static void item_set(u8 i, u8 v)
{
    u8 mask = 1 << (i & 7);
    if (v) item_bits[i >> 3] |= mask;
    else   item_bits[i >> 3] &= (u8)~mask;
}

static void flag_set(u8 i, u8 v)
{
    u8 mask = 1 << (i & 7);
    if (v) flag_bits[i >> 3] |= mask;
    else   flag_bits[i >> 3] &= (u8)~mask;
}

/* --- Conditions -------------------------------------------------------- */

static u8 eval_atom(u8 op, u8 a0, u8 a1, u8 a2)
{
    u8 v;
    switch (op) {
    case OP_FLAG_SET: return flag_get(a0);
    case OP_FLAG_CLR: return (u8)!flag_get(a0);
    case OP_HAS_ITEM: return item_get(a0);
    case OP_NO_ITEM:  return (u8)!item_get(a0);
    case OP_STAT_CMP:
        v = stat_val[a0];
        switch (a1) {
        case CMP_EQ: return (u8)(v == a2);
        case CMP_NE: return (u8)(v != a2);
        case CMP_LT: return (u8)(v <  a2);
        case CMP_LE: return (u8)(v <= a2);
        case CMP_GT: return (u8)(v >  a2);
        case CMP_GE: return (u8)(v >= a2);
        }
        return 0;
    }
    return 0;
}

u8 state_eval_cond(void)
{
    u8 n = b_u8();
    u8 conn = b_u8();     /* 0=AND, 1=OR */
    u8 result, i, av, op, a0, a1, a2;

    if (n == 0)
        return 1;

    result = (conn == 0) ? 1 : 0;   /* AND part de vrai, OR de faux */
    for (i = 0; i < n; ++i) {
        op = b_u8(); a0 = b_u8(); a1 = b_u8(); a2 = b_u8();
        av = eval_atom(op, a0, a1, a2);
        if (conn == 0) result = (u8)(result & av);
        else           result = (u8)(result | av);
    }
    return result;
}

/* --- Effets ------------------------------------------------------------ */

static void stat_clamp_set(u8 idx, int nv)
{
    if (nv < (int)stat_min[idx]) nv = stat_min[idx];
    if (nv > (int)stat_max[idx]) nv = stat_max[idx];
    stat_val[idx] = (u8)nv;
}

void state_skip_effects(void)
{
    u8 n = b_u8();
    u8 i, na;
    for (i = 0; i < n; ++i) {
        na = b_u8();                            /* garde : n_atoms */
        (void)b_u8();                           /* connective */
        b_seek((u16)(b_tell() + (u16)na * 4));  /* atomes */
        b_seek((u16)(b_tell() + 4));            /* op + 3 operandes */
    }
}

u16 state_apply_effects(void)
{
    u8 n = b_u8();
    u8 i, op, a0, a1, ok;
    u16 target = NO_GOTO;

    for (i = 0; i < n; ++i) {
        ok = state_eval_cond();                 /* garde de l'effet */
        op = b_u8(); a0 = b_u8(); a1 = b_u8(); (void)b_u8();  /* a2 inutilisé */
        if (!ok)
            continue;                           /* garde fausse -> effet ignore */
        switch (op) {
        case OP_SET_FLAG:  flag_set(a0, 1); break;
        case OP_CLR_FLAG:  flag_set(a0, 0); break;
        case OP_TOG_FLAG:  flag_set(a0, (u8)!flag_get(a0)); break;
        case OP_GIVE_ITEM: item_set(a0, 1); break;
        case OP_TAKE_ITEM: item_set(a0, 0); break;
        case OP_STAT_ADD:  stat_clamp_set(a0, (int)stat_val[a0] + a1); break;
        case OP_STAT_SUB:  stat_clamp_set(a0, (int)stat_val[a0] - a1); break;
        case OP_STAT_SET:  stat_clamp_set(a0, (int)a1); break;
        case OP_SOUND:
            snd_play(a0);
            break;
        case OP_SCORE_ADD:
            if (g_score_on) g_score += a0;
            break;
        case OP_STAT_MAX:                   /* ~ restore STAT : au max courant */
            stat_val[a0] = stat_max[a0];
            break;
        case OP_STAT_SETMAX:                /* ~ setmax STAT N : fixe le max */
            stat_max[a0] = a1;
            if (stat_val[a0] > a1) stat_val[a0] = a1;
            break;
        case OP_GOTO:
            target = a0 | ((u16)a1 << 8);
            return target;   /* saut immédiat (spec §5.2) */
        }
    }
    return target;
}

File: player/webng/src/app/engine/state.ts
// state.ts -- etat runtime du joueur + evaluation conditions/effets, miroir
// de state.c. Fonctionne sur les structures decodees par section.ts
// (CondBlock/EffectList) plutot que sur un curseur d'octets -- cf. l'entete
// de section.ts pour pourquoi.

import type { CondBlock, EffectList } from "./section";
import type { StoryData } from "./story";
import { flagDefaultBit, itemDefaultBit } from "./story";
import {
  CMP_EQ,
  CMP_GE,
  CMP_GT,
  CMP_LE,
  CMP_LT,
  CMP_NE,
  NO_GOTO,
  OP_CLR_FLAG,
  OP_FLAG_CLR,
  OP_FLAG_SET,
  OP_GIVE_ITEM,
  OP_GOTO,
  OP_HAS_ITEM,
  OP_NO_ITEM,
  OP_SCORE_ADD,
  OP_SET_FLAG,
  OP_SOUND,
  OP_STAT_ADD,
  OP_STAT_CMP,
  OP_STAT_MAX,
  OP_STAT_SET,
  OP_STAT_SETMAX,
  OP_STAT_SUB,
  OP_TAKE_ITEM,
  OP_TOG_FLAG,
} from "./format";

export class GameState {
  statVal: number[];
  statMax: number[]; // max COURANT (mutable via ~ setmax, cf. OP_STAT_SETMAX)
  itemBits: Uint8Array;
  flagBits: Uint8Array;
  score = 0;
  moves = 0;
  private readonly story: StoryData;

  constructor(story: StoryData) {
    this.story = story;
    this.statVal = story.statInit.slice();
    this.statMax = story.statMaxDef.slice();
    this.itemBits = new Uint8Array(Math.ceil(story.nItems / 8));
    this.flagBits = new Uint8Array(Math.ceil(story.nFlags / 8));
    for (let i = 0; i < story.nItems; ++i) this.setItemBit(i, itemDefaultBit(story, i));
    for (let i = 0; i < story.nFlags; ++i) this.setFlagBit(i, flagDefaultBit(story, i));
  }

  itemGet(i: number): boolean {
    return ((this.itemBits[i >> 3] >> (i & 7)) & 1) !== 0;
  }
  flagGet(i: number): boolean {
    return ((this.flagBits[i >> 3] >> (i & 7)) & 1) !== 0;
  }
  private setItemBit(i: number, v: number): void {
    const mask = 1 << (i & 7);
    if (v) this.itemBits[i >> 3] |= mask;
    else this.itemBits[i >> 3] &= ~mask & 0xff;
  }
  private setFlagBit(i: number, v: number): void {
    const mask = 1 << (i & 7);
    if (v) this.flagBits[i >> 3] |= mask;
    else this.flagBits[i >> 3] &= ~mask & 0xff;
  }

  /** Remet a 0 les flags LOCAUX ([localBase, nFlags)) -- cf.
   * state_clear_locals(), appele a chaque changement de fichier/chapitre. */
  clearLocals(): void {
    for (let i = this.story.localBase; i < this.story.nFlags; ++i) this.setFlagBit(i, 0);
  }

  /** Sommes des modificateurs de combat de tous les objets PORTES. */
  gearAtk(): number {
    return this.sumGear(this.story.itemAtk);
  }
  gearDmg(): number {
    return this.sumGear(this.story.itemDmg);
  }
  gearArmor(): number {
    return this.sumGear(this.story.itemArmor);
  }
  private sumGear(mods: number[]): number {
    let s = 0;
    for (let i = 0; i < this.story.nItems; ++i) if (this.itemGet(i)) s += mods[i];
    return s;
  }

  private evalAtom(op: number, a0: number, a1: number, a2: number): boolean {
    switch (op) {
      case OP_FLAG_SET:
        return this.flagGet(a0);
      case OP_FLAG_CLR:
        return !this.flagGet(a0);
      case OP_HAS_ITEM:
        return this.itemGet(a0);
      case OP_NO_ITEM:
        return !this.itemGet(a0);
      case OP_STAT_CMP: {
        const v = this.statVal[a0];
        switch (a1) {
          case CMP_EQ:
            return v === a2;
          case CMP_NE:
            return v !== a2;
          case CMP_LT:
            return v < a2;
          case CMP_LE:
            return v <= a2;
          case CMP_GT:
            return v > a2;
          case CMP_GE:
            return v >= a2;
        }
        return false;
      }
    }
    return false;
  }

  evalCond(cond: CondBlock): boolean {
    if (cond.clauses.length === 0) return true;
    return cond.clauses.some((clause) =>
      clause.every((a) => this.evalAtom(a.op, a.a0, a.a1, a.a2)),
    );
  }

  private clampSet(idx: number, nv: number): void {
    if (nv < this.story.statMin[idx]) nv = this.story.statMin[idx];
    if (nv > this.statMax[idx]) nv = this.statMax[idx];
    this.statVal[idx] = nv;
  }

  /** Applique un effect_list ; renvoie l'index de section d'un GOTO
   * rencontre (short-circuit, comme state_apply_effects), ou NO_GOTO.
   * `playSound` est optionnel : NO-OP si non fourni (ex. lors d'une reprise
   * de sauvegarde ou d'un test hors navigateur). */
  applyEffects(list: EffectList, playSound?: (id: number) => void): number {
    let target = NO_GOTO;
    for (const eff of list) {
      if (!this.evalCond(eff.guard)) continue;
      switch (eff.op) {
        case OP_SET_FLAG:
          this.setFlagBit(eff.a0, 1);
          break;
        case OP_CLR_FLAG:
          this.setFlagBit(eff.a0, 0);
          break;
        case OP_TOG_FLAG:
          this.setFlagBit(eff.a0, this.flagGet(eff.a0) ? 0 : 1);
          break;
        case OP_GIVE_ITEM:
          this.setItemBit(eff.a0, 1);
          break;
        case OP_TAKE_ITEM:
          this.setItemBit(eff.a0, 0);
          break;
        case OP_STAT_ADD:
          this.clampSet(eff.a0, this.statVal[eff.a0] + eff.a1);
          break;
        case OP_STAT_SUB:
          this.clampSet(eff.a0, this.statVal[eff.a0] - eff.a1);
          break;
        case OP_STAT_SET:
          this.clampSet(eff.a0, eff.a1);
          break;
        case OP_SOUND:
          playSound?.(eff.a0);
          break;
        case OP_SCORE_ADD:
          if (this.story.scoreOn) this.score += eff.a0;
          break;
        case OP_STAT_MAX: // ~ restore STAT : au max courant
          this.statVal[eff.a0] = this.statMax[eff.a0];
          break;
        case OP_STAT_SETMAX: // ~ setmax STAT N : fixe le max
          this.statMax[eff.a0] = eff.a1;
          if (this.statVal[eff.a0] > eff.a1) this.statVal[eff.a0] = eff.a1;
          break;
        case OP_GOTO:
          return eff.a0 | (eff.a1 << 8); // saut immediat, effets suivants ignores
      }
    }
    return target;
  }
}
